### PR TITLE
feat: MKLDNN LSTM unidirectional/bidirectional inference support

### PR DESCRIPTION
**Pull request by benjamim93**
## What changes were proposed in this pull request?
Enable MKLDNN LSTM unidirectional/bidirectional inference:
1. Unidirectional Left2Right
2. Unidirectional Right2Left
3. Bidirectional Concat
4. Bidirectional Sum
(Please fill in changes proposed in this patch)

## How was this patch tested?
Created layers in both MKLDNN and BLAS with following parameters:
1. Unidirectional Left2Right and Right2Left
    Sequence Length = 7
    Batch Size = 2
    Input Size = 3
    Hidden Size = 5
    Output activation is Tanh
2. Bidirectional Sum and Concat
    Sequence Length = 3
    Batch Size = 2
    Input Size = 3
    Hidden Size = 5
    Output activation is Tanh

Inputs, weights and biases were randomly generated (uniform distribution).
Results were compared using Equivalent.nearequals()

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If it is possible, please attach a screenshot; otherwise, remove this)

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LSTM2.scala [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+import scala.reflect.ClassTag
+
+/**
+ * Long Short Term Memory architecture similar to nn/LSTM but without Dropout
+ * It is used to verify the inference results of mkldnn/LSTM
+ */
+// @SerialVersionUID(- 8176191554025511686L)
+class LSTM2[T : ClassTag] (
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LSTM2.scala]: why LSTM2? we need a better name

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryData.scala [hunk 2]
```diff
@@ -278,4 +278,18 @@ private[mkldnn] object MemoryData {
     memory.setPrimitiveDescription(memoryPrimDesc)
     memory
   }
+
+  def operationWant2(primDesc: Long, queryType: Int, index: Int): NativeData = {
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryData.scala]: why operationWant2? need a better name

**Comment by wzhongyuan** [comment]
No need to create a separate LSTM, we can reuse the existing one for DNN implementation, probably we just need to set scale to `true`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/LSTM.scala [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    updateWithNewTensor(updateOutputTensors, 0, input)
+
+    MklDnnOps.streamSubmit(runtime.stream, 1, updateOutputPrimitives, updateOutputPrimitives.length,
+      updateOutputMemoryPrimitives, updateOutputTensors)
+
+    output
+  }
+
+  override private[mkldnn] def initBwdPrimitives(grad: Array[MemoryData], phase: Phase) = {
+    _gradInputFormats = grad.clone()
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/LSTM.scala]: If this layer does not support backward, you should throw exception if calling backward.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/LSTM.scala [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+import com.intel.analytics.bigdl.mkl._
+import com.intel.analytics.bigdl.nn.{InitializationMethod, RandomUniform, VariableFormat}
+import com.intel.analytics.bigdl.nn.abstractnn.{Activity, Initializable}
+import com.intel.analytics.bigdl.nn.mkldnn.Phase.InferencePhase
+import com.intel.analytics.bigdl.tensor.Tensor
+
+import scala.collection.mutable.ArrayBuffer
+
+class LSTM(
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/LSTM.scala]: Add comments for all parameters

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/LSTM.scala [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+  val inputSize: Int,
+  val hiddenSize: Int,
+  val f: Int,
+  val direction: Int,
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/LSTM.scala]: add description

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LSTM2.scala [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+  override def reset(): Unit = {
+    super.reset()
+    cell.reset()
+  }
+
+
+  override def toString: String = s"LSTM2($inputSize, $hiddenSize)"
+}
+
+object LSTM2 {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LSTM2.scala]: This layer is just for unit tests, right?
If so, you should move it to test package.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryData.scala (hunk 2 above)
Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryData.scala]: Yes, the method has been renamed operationWantWithIndex.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LSTM2.scala (hunk 1 above)
Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LSTM2.scala]: LSTMSpec has been modified and LSTM2 is not used anymore. Old LSTM is enough.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+  private[mkldnn] var weight: TensorMMap = _
+  private[mkldnn] var weight_i: TensorMMap = _
+  private[mkldnn] var bias: TensorMMap = _
+  private[mkldnn] var src_i: TensorMMap = _
+  private[mkldnn] var dst_i: TensorMMap = _
+
+  if(layers > 1) {
+    require(inputSize == hiddenSize,
+      "If layers of LSTM is more than 1," +
+        " the input size and the hidden size should equal")
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala]: print inputSize and outputSize if not meet requirement

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+          */
+        weight = new TensorMMap(Array(common_n_layers, 1, inputSize, ngates, hiddenSize))
+        weight_i = new TensorMMap(Array(common_n_layers, 1, hiddenSize, ngates, hiddenSize))
+        bias = new TensorMMap(Array(common_n_layers, 1, ngates, hiddenSize))
+
+        {
+          val stdv = 1.0 / math.sqrt(hiddenSize)
+          val wInit: InitializationMethod = RandomUniform(-stdv, stdv)
+          val bInit: InitializationMethod = Zeros
+          setInitMethod(wInit, bInit)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala]: should not init weight here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+      case Direction.UnidirectionalLeft2Right
+           | Direction.UnidirectionalRight2Left =>
+        /**
+          * Gate order matching between MKLDNN LSTM and nn/LSTM:
+          * MKLDNN Gate 1 -> nn/LSTM Gate 1 (input gate)
+          * MKLDNN Gate 2 -> nn/LSTM Gate 3 (forget gate)
+          * MKLDNN Gate 3 -> nn/LSTM Gate 2 (hidden)
+          * MKLDNN Gate 4 -> nn/LSTM Gate 4 (output gate)
+          */
+        weight = new TensorMMap(Array(common_n_layers, 1, inputSize, ngates, hiddenSize))
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala]: You must create weight when creating RNN instance.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryData.scala [hunk 10]
```diff
@@ -278,4 +278,18 @@ private[mkldnn] object MemoryData {
     memory.setPrimitiveDescription(memoryPrimDesc)
     memory
   }
+
+  def operationWantWithIndex(primDesc: Long, queryType: Int, index: Int): NativeData = {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryData.scala]: Please not add new method, but enhance `operationWant` with parameter `index`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+    }
+
+    updateWithNewTensor(updateOutputTensors, 0, input)
+
+    MklDnnOps.streamSubmit(runtime.stream, 1, updateOutputPrimitives, updateOutputPrimitives.length,
+      updateOutputMemoryPrimitives, updateOutputTensors)
+
+    output
+  }
+
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala]: Add some release code to release internal tensor.

**Comment by zhangxiaoli73** [comment]
Please override model release() parameters() zeroGradParameters()

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+    if (initBias == null) {
+      biasInitMethod.init(bias.dense, VariableFormat.Default)
+    } else {
+      bias.dense.copy(initBias)
+    }
+  }
+
+  private def initMemoryDescs(inputs: Array[MemoryData]) = {
+    val(inputShape, inputLayout) = inputs(0).layout match {
+      case Memory.Format.tnc => /* tnc */
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala]: input format may be not `tnc`.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+    val kind = if (!isTraining()) {
+      PropKind.ForwardInference
+    } else {
+      throw new UnsupportedOperationException("Not support training")
+    }
+
+    val rnnCellDesc = mode match {
+      case AlgKind.VanillaLstm =>
+        MklDnn.RNNCellDescInit(AlgKind.VanillaLstm, f, flags, alpha, clipping)
+      case _ => throw new UnsupportedOperationException("Not support such cell")
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala]: print mode

Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala]: Several throwing exceptions have been updated to print more info. Input format check added. More members were added to release code.

**Comment by benjamim93** [comment]
BigDL-PRVN-lite - Build # 636 - Successful:

Check console output at http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/636/ to view the results.

The pull request: https://github.com/intel-analytics/spark-dl/pull/2806 Successful

**Pull request merged**